includes/shell/logger.hpp

```diff
@@ -2,66 +2,116 @@
 
 #include "shell/settings.hpp"
 #include "shell/print.hpp"
+#include "shell/util.hpp"
 
-#define LOG_INFO(format, ...)  shell::Logger::getInstance().write(shell::Logger::LogLevel::Info, format, ##__VA_ARGS__)
-#define LOG_ERROR(format, ...) shell::Logger::getInstance().write(shell::Logger::LogLevel::Error, format, ##__VA_ARGS__)
-#define LOG_WARNING(format, ...)                                                                                       \
-	shell::Logger::getInstance().write(shell::Logger::LogLevel::Warning, format, ##__VA_ARGS__)
+#include <sstream>
+
+#define LOG_INFO(format, ...)    shell::SLogger::getInstance().write(shell::LogLevel::Info, format, ##__VA_ARGS__)
+#define LOG_ERROR(format, ...)   shell::SLogger::getInstance().write(shell::LogLevel::Error, format, ##__VA_ARGS__)
+#define LOG_WARNING(format, ...) shell::SLogger::getInstance().write(shell::LogLevel::Warning, format, ##__VA_ARGS__)
 
 #ifdef DEBUG
-// todo only print `[DEBUG]` at start of line (i.e. handle flushing)
-#define LOG_DEBUG(format, ...) shell::Logger::getInstance().write(shell::Logger::LogLevel::Debug, format, ##__VA_ARGS__)
+#define LOG_DEBUG(format, ...) shell::SLogger::getInstance().write(shell::LogLevel::Debug, format, ##__VA_ARGS__)
 #else
 #define LOG_DEBUG(format, ...)
 #endif
 
 namespace shell {
 
 /*!
- * @brief Manages Logging.
+ * @brief -.
+ */
+enum class LogLevel { Info, Warning, Error, Debug };
+
+/*!
+ * @brief Logger baseclass can only be instantiated by derived class.
  */
 class Logger {
-	Logger() = default;
+protected:
+	const LogLevel g_logLevel; /*!< Loglevel set at compiletime. */
 
-public:
 	/*!
 	 * @brief -.
 	 */
-	enum class LogLevel { Info, Warning, Error, Debug };
+	Logger();
 
 	/*!
-	 * @brief Writes message like printf but with loglevel prefix.
+	 * @brief Gets loglevel prefix as string.
+	 */
+	static const string getLogLevelPrefix(LogLevel logLevel);
+};
+
+/*!
+ * @brief Singleton Logging class.
+ * @note This implementation is not thread safe.
+ */
+class SLogger : public Logger {
```

Comment:
Can we just call it SingletonLogger and BufferedLogger for clarity?

Reply:
Hahaha yeah i was going to rework it anyways  👍🏽